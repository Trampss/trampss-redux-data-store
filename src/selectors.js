import { at } from 'lodash'

export const getState = path => prefix => (state) => {
  let subState = state
  if (path !== undefined) subState = at(state, path)[0]

  return subState[prefix]
}

const getFactory = key => path => prefix => state => getState(path)(prefix)(state)[key]
export const getKeys = getFactory('keys')
export const getAsArray = getFactory('array')
export const getNb = getFactory('nb')
export const isInitialized = getFactory('initialized')

const getDatas = getFactory('datas')
export const get = path => prefix => state => (keys) => {
  const datas = getDatas(path)(prefix)(state)
  // All datas
  if (!keys) return datas
  // By keys
  if (Array.isArray(keys)) return keys.map(k => datas[k])
  // By key
  return datas[keys]
}

