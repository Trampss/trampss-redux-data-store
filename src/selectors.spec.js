/* eslint-env jest */
import { getKeys, getLength, getState, get, getBy, getAsArray, isInitialized } from './selectors'

const Element = code => ({ code, some: 'other', infos: code })
const SubElement = code => subCode => ({ ...Element(code), sub: { subCode } })

const path = 'api.raw'
const prefix = 'testPrefix'
const subState = {
  testPrefix: {
    data: {
      elm2: Element('elm2'),
      elm1: Element('elm1'),
      elm3: Element('elm3'),
      elm4: SubElement('elm4')('subelm4'),
      elm5: SubElement('elm5')('subelm5'),
    },
    keys: ['elm2', 'elm1', 'elm3', 'elm4', 'elm5'],
    array: [Element('elm2'), Element('elm1'), Element('elm3'), SubElement('elm4')('subelm4'), SubElement('elm5')('subelm5')],
    initialized: true,
  },
}
const state = {
  api: {
    raw: {
      ...subState,
    },
  },
}

describe('selectors', () => {
  it('should retrieve state', () => expect(getState(path)(prefix)(state)).toMatchSnapshot())
  it('should retrieve state, without path set', () => expect(getState()(prefix)(subState)).toMatchSnapshot())
  it('should retrieve keys', () => expect(getKeys(path)(prefix)(state)).toMatchSnapshot())
  it('should retrieve length', () => expect(getLength(path)(prefix)(state)).toMatchSnapshot())
  it('should retrieve initialized value', () => expect(isInitialized(path)(prefix)(state)).toMatchSnapshot())
  it('should retrieve all data, without key', () => expect(get(path)(prefix)()(state)).toMatchSnapshot())
  it('should retrieve data by ids, with an array of keys', () => expect(get(path)(prefix)(['elm1', 'elm3'])(state)).toMatchSnapshot())
  it('should retrieve data by id, with a single key', () => expect(get(path)(prefix)('elm3')(state)).toMatchSnapshot())
  it('should retrieve all data, without key to array', () => expect(getAsArray(path)(prefix)(state)).toMatchSnapshot())
  it('should retrieve all data, validate (path, values), with simple path and simple value', () => expect(getBy(path)(prefix)('code', 'elm2')(state)).toMatchSnapshot())
  it('should retrieve all data, validate (path, values), with sub path and simple value', () => expect(getBy(path)(prefix)('sub.subCode', 'subelm4')(state)).toMatchSnapshot())
  it('should retrieve all data, validate (path, values), with simple path and multi value', () => expect(getBy(path)(prefix)('code', ['elm1', 'elm4', 'subelm5'])(state)).toMatchSnapshot())
  it('should retrieve all data, validate (path, values), with sub path and multi value', () => expect(getBy(path)(prefix)('sub.subCode', ['subelm4', 'subelm5', 'elm1'])(state)).toMatchSnapshot())
  it('should retrieve all data, validate (path, values), with wrong path', () => expect(getBy(path)(prefix)('sub.wrong', 'subelm4')(state)).toMatchSnapshot())
  it('should retrieve all data, validate (path, values), when not find value', () => expect(getBy(path)(prefix)('sub.wrong', 'subelm4')(state)).toMatchSnapshot())
})
