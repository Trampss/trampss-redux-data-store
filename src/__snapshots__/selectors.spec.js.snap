// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors bugs should works with empty path (not undefined) for first level reducer 1`] = `"firstLevel"`;

exports[`selectors should retrieve all data, validate (path, values), when not find value 1`] = `Array []`;

exports[`selectors should retrieve all data, validate (path, values), with simple path and multi value 1`] = `
Array [
  Object {
    "code": "elm1",
    "infos": "elm1",
    "some": "other",
  },
  Object {
    "code": "elm4",
    "infos": "elm4",
    "some": "other",
    "sub": Object {
      "subCode": "subelm4",
    },
  },
]
`;

exports[`selectors should retrieve all data, validate (path, values), with simple path and simple value 1`] = `
Array [
  Object {
    "code": "elm2",
    "infos": "elm2",
    "some": "other",
  },
]
`;

exports[`selectors should retrieve all data, validate (path, values), with sub path and multi value 1`] = `
Array [
  Object {
    "code": "elm4",
    "infos": "elm4",
    "some": "other",
    "sub": Object {
      "subCode": "subelm4",
    },
  },
  Object {
    "code": "elm5",
    "infos": "elm5",
    "some": "other",
    "sub": Object {
      "subCode": "subelm5",
    },
  },
]
`;

exports[`selectors should retrieve all data, validate (path, values), with sub path and simple value 1`] = `
Array [
  Object {
    "code": "elm4",
    "infos": "elm4",
    "some": "other",
    "sub": Object {
      "subCode": "subelm4",
    },
  },
]
`;

exports[`selectors should retrieve all data, validate (path, values), with wrong path 1`] = `Array []`;

exports[`selectors should retrieve all data, without key 1`] = `
Object {
  "elm1": Object {
    "code": "elm1",
    "infos": "elm1",
    "some": "other",
  },
  "elm2": Object {
    "code": "elm2",
    "infos": "elm2",
    "some": "other",
  },
  "elm3": Object {
    "code": "elm3",
    "infos": "elm3",
    "some": "other",
  },
  "elm4": Object {
    "code": "elm4",
    "infos": "elm4",
    "some": "other",
    "sub": Object {
      "subCode": "subelm4",
    },
  },
  "elm5": Object {
    "code": "elm5",
    "infos": "elm5",
    "some": "other",
    "sub": Object {
      "subCode": "subelm5",
    },
  },
}
`;

exports[`selectors should retrieve all data, without key to array 1`] = `
Array [
  Object {
    "code": "elm2",
    "infos": "elm2",
    "some": "other",
  },
  Object {
    "code": "elm1",
    "infos": "elm1",
    "some": "other",
  },
  Object {
    "code": "elm3",
    "infos": "elm3",
    "some": "other",
  },
  Object {
    "code": "elm4",
    "infos": "elm4",
    "some": "other",
    "sub": Object {
      "subCode": "subelm4",
    },
  },
  Object {
    "code": "elm5",
    "infos": "elm5",
    "some": "other",
    "sub": Object {
      "subCode": "subelm5",
    },
  },
]
`;

exports[`selectors should retrieve data by id, with a single key 1`] = `
Object {
  "code": "elm3",
  "infos": "elm3",
  "some": "other",
}
`;

exports[`selectors should retrieve data by ids, with an array of keys 1`] = `
Array [
  Object {
    "code": "elm1",
    "infos": "elm1",
    "some": "other",
  },
  Object {
    "code": "elm3",
    "infos": "elm3",
    "some": "other",
  },
]
`;

exports[`selectors should retrieve initialized value 1`] = `true`;

exports[`selectors should retrieve keys 1`] = `
Array [
  "elm2",
  "elm1",
  "elm3",
  "elm4",
  "elm5",
]
`;

exports[`selectors should retrieve length 1`] = `5`;

exports[`selectors should retrieve state 1`] = `
Object {
  "array": Array [
    Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    Object {
      "code": "elm1",
      "infos": "elm1",
      "some": "other",
    },
    Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other",
    },
    Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
      "sub": Object {
        "subCode": "subelm4",
      },
    },
    Object {
      "code": "elm5",
      "infos": "elm5",
      "some": "other",
      "sub": Object {
        "subCode": "subelm5",
      },
    },
  ],
  "data": Object {
    "elm1": Object {
      "code": "elm1",
      "infos": "elm1",
      "some": "other",
    },
    "elm2": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "elm3": Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other",
    },
    "elm4": Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
      "sub": Object {
        "subCode": "subelm4",
      },
    },
    "elm5": Object {
      "code": "elm5",
      "infos": "elm5",
      "some": "other",
      "sub": Object {
        "subCode": "subelm5",
      },
    },
  },
  "initialized": true,
  "keys": Array [
    "elm2",
    "elm1",
    "elm3",
    "elm4",
    "elm5",
  ],
}
`;

exports[`selectors should retrieve state, without path set 1`] = `
Object {
  "array": Array [
    Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    Object {
      "code": "elm1",
      "infos": "elm1",
      "some": "other",
    },
    Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other",
    },
    Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
      "sub": Object {
        "subCode": "subelm4",
      },
    },
    Object {
      "code": "elm5",
      "infos": "elm5",
      "some": "other",
      "sub": Object {
        "subCode": "subelm5",
      },
    },
  ],
  "data": Object {
    "elm1": Object {
      "code": "elm1",
      "infos": "elm1",
      "some": "other",
    },
    "elm2": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "elm3": Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other",
    },
    "elm4": Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
      "sub": Object {
        "subCode": "subelm4",
      },
    },
    "elm5": Object {
      "code": "elm5",
      "infos": "elm5",
      "some": "other",
      "sub": Object {
        "subCode": "subelm5",
      },
    },
  },
  "initialized": true,
  "keys": Array [
    "elm2",
    "elm1",
    "elm3",
    "elm4",
    "elm5",
  ],
}
`;
